### ISSUE TITLE
Occupancy masks for geometric transforms 

### ISSUE THREAD

**Issue by vadimkantorov** [kornia/kornia#1000]
Consideer an equivariance loss scenario: we have a source image `src`, we get `view1 = k1(src); view2 = k2(src)`. Now we'd like to put on some loss on corresponding locations in `view1` and `view2`.

For most of geometric augmentations, not all pixel locations in `view1` would have correspondences in `view2`, and loss must be computed only for valid locations, i.e. we should not compute loss for black pixels in rotated images in https://github.com/kornia/kornia-examples/blob/master/data_augmenation_segmentation.ipynb 

In this notebook is advised to use `k1(img, k1._params)` for repeating the sampled transformation on another input. Is there an easy way of performing an inverse transform given `_params`?

I think the correct expression for mask of valid locations in `view` is given by `mask_valid_1 = k1(k2(k2(ones_like(src), k2._params), k2._params, INVERSE = True), k1._params)`, but one needs an easy way of doing an inverse transform.

(related https://github.com/kornia/kornia/issues/476#issuecomment-833147695)

As a workaround, I'm not manually using `return_transform`, and doing `warp_affine` with inverse transforms

**Comment by edgarriba** [comment]
@vadimkantorov doesn't work the `return_transform` flag for you ?

**Comment by vadimkantorov** [comment]
It works, but it breaks the API elegance, I end up mixing  `k1(img, k2._params)` and `K.warp_affine(img, t2.inverse(), img.shape[-2:])`

**Comment by edgarriba** [comment]
absolutely - adding a `return_inverse` flag could be definitely quite neat. /cc @ducha-aiki @shijianjian @lferraz thoughts ?

**Comment by vadimkantorov** [comment]
Other solution is to give up on `_params` and compose final transforms manually with matrix products and doing a single final `K.warp_affine` - maybe still a good thing to do from perf standpoint.

But the scenario of equivariance losses is definitely important for self-supervised tasks

**Comment by vadimkantorov** [comment]
Compare:
```python
# mixed styles
mask1 = rotate1(K.warp_affine(this.rotate2(mask, this.rotate2._params), T2.inverse(), spatial_size), k1._params)
mask2 = rotate2(K.warp_affine(this.rotate1(mask, this.rotate1._params), T1.inverse(), spatial_size), k2._params)

# only warp affine
mask1 = K.warp_affine(K.warp_affine(K.warp_affine(mask, T2, view2.shape[-2:]), T2.inverse(), spatial_size), T1, view1.shape[-2:])  
mask2 = K.warp_affine(K.warp_affine(K.warp_affine(mask, T1, view1.shape[-2:]), T1.inverse(), spatial_size), T2, view2.shape[-2:])
```

**Comment by ducha-aiki** [comment]
I believe, that flag return_inverse is the best option.

**Comment by vadimkantorov** [comment]
Whatever the decision, I propose to have some recommended way in the docs (be it wrap_affine-only style, or some mixed style).

Some examples of recent pipelines are: https://github.com/snap-research/articulated-animation/blob/66979bea61ddd8166c13008d5b603740ea105cd5/modules/model.py#L82, https://github.com/facebookresearch/dino/blob/main/main_dino.py#L415

**Comment by shijianjian** [comment]
From an API level, it may make sense to have an additional 'apply_inverse' for each augmentation. Then, we can avoid access the '_params':

'''python
aug = RandomRotation(360.)
img_rot = aug(img)
mask_rot = aug(mask, aug._params)

inversed_img = aug.apply_inverse(some_img)
'''

**Comment by vadimkantorov** [comment]
Well, I don't mind explicitly typing `aug._params` to highlight that saved params are used. `apply_inverse` is more hideous in this respect

**Comment by shijianjian** [comment]
HI @vadimkantorov, are you willing to send a PR for your proposal?

**Comment by vadimkantorov** [comment]
realistically - not likely :(

**Comment by edgarriba** [comment]
merged in #1013
